### PR TITLE
Fix view bugs where views create conflicting metric streams at runtime

### PR DESCRIPTION
**Pull request by alanwest**
This PR reopens #3074. It is an alternative to #3006.

In this one, views that create conflicts at runtime are ignored.

### CODE REVIEW

File: test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs [hunk 1]
```diff
@@ -469,6 +469,244 @@ long GetAggregatedValue(Metric metric)
...
+        [Fact]
+        public void DuplicateInstrumentRegistration_WithViews_TwoIdenticalInstruments_TwoViews_SameRename()
+        {
+            var exportedItems = new List<Metric>();
+
+            using var meter = new Meter($"{Utils.GetCurrentMethodName()}");
+            var meterProviderBuilder = Sdk.CreateMeterProviderBuilder()
+                .AddMeter(meter.Name)
+                .AddView("name", "newname")
+                .AddView("name", "newname")
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:481]: Would it be a good idea to throw here (e.g. it's possible to detect that two `View`s are using the same `newname` before we see any instruments).

Comment (alanwest) [reply test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:481]: Yes, agreed. @cijothomas and I have discussed this a bit. There are definitely some improvements we can make at view registration time and fail-fast when we know for certain a view will cause conflicts.

It's namely our `AddView(Func<Instrument, MetricStreamConfig>)` overload where we won't be able to detect a conflict at registration time. I think I will change this test to use that overload instead to achieve a similar result.

I'd like to consider cases where we can fail fast at registration in a different PR.

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs [hunk 2]
```diff
@@ -42,7 +42,7 @@ internal Metric AddMetricWithNoViews(Instrument instrument)
             var meterVersion = instrument.Meter.Version;
             var metricName = instrument.Name;
             var metricStreamName = $"{meterName}.{meterVersion}.{metricName}";
-            var instrumentIdentity = new InstrumentIdentity(instrument.Meter, metricName, instrument.Unit, instrument.Description, instrument.GetType());
+            var instrumentIdentity = new InstrumentIdentity(instrument.Meter, metricName, instrument.Unit, instrument.Description, instrument.GetType(), false);
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricReaderExt.cs:45]: nit: I personally would pass like `derivedFromView: false`. Just helps with readability to throw the parameter name on there when the thing you are passing doesn't have name (typically `true`, `false`, & `null`).

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs [hunk 3]
```diff
@@ -119,7 +119,7 @@ internal List<Metric> AddMetricsListWithViews(Instrument instrument, List<Metric
                     var metricName = metricStreamConfig?.Name ?? instrument.Name;
                     var metricStreamName = $"{meterName}.{meterVersion}.{metricName}";
                     var metricDescription = metricStreamConfig?.Description ?? instrument.Description;
-                    var instrumentIdentity = new InstrumentIdentity(instrument.Meter, metricName, instrument.Unit, metricDescription, instrument.GetType());
+                    var instrumentIdentity = new InstrumentIdentity(instrument.Meter, metricName, instrument.Unit, metricDescription, instrument.GetType(), metricStreamConfig != null);
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricReaderExt.cs:122]: nit: Same feedback here eg `derivedFromView: metricStreamConfig != null`

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs [hunk 4]
```diff
@@ -133,20 +133,22 @@ internal List<Metric> AddMetricsListWithViews(Instrument instrument, List<Metric
...
                             metricName,
                             meterName,
                             "Metric instrument has the same name as an existing one but differs by description, unit, or instrument type. Measurements from this instrument will still be exported but may result in conflicts.",
                             "Either change the name of the instrument or use MeterProviderBuilder.AddView to resolve the conflict.");
+
+                        if (metrics.Count == 0)
+                        {
+                            metrics.Add(existingMetric);
+                        }
+
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricReaderExt.cs:147]: The else part is where we ignore the view and move on - we should do `OpenTelemetrySdkEventSource.Log.MetricViewIgnored` here.

File: test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs [hunk 5]
```diff
@@ -469,6 +469,244 @@ long GetAggregatedValue(Metric metric)
             }
         }
 
+        [Fact]
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:472]: Nit: move these tests to the MetricViewTests.cs

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs [hunk 6]
```diff
@@ -133,20 +133,22 @@ internal List<Metric> AddMetricsListWithViews(Instrument instrument, List<Metric
...
+                        continue;
                     }
 
+                    // TODO: I think this check needs to be moved up...
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricReaderExt.cs:151]: yes.

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs [hunk 7]
```diff
@@ -133,20 +133,22 @@ internal List<Metric> AddMetricsListWithViews(Instrument instrument, List<Metric
...
 
                     if (this.instrumentIdentityToMetric.TryGetValue(instrumentIdentity, out var existingMetric))
-                    {
-                        metrics.Add(existingMetric);
-                        continue;
-                    }
-
-                    if (this.metricStreamNames.Contains(metricStreamName))
                     {
                         OpenTelemetrySdkEventSource.Log.DuplicateMetricInstrument(
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricReaderExt.cs:137]: This log must only be inside the if (metrics.Count==0) part.

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs [hunk 8]
```diff
@@ -133,20 +133,22 @@ internal List<Metric> AddMetricsListWithViews(Instrument instrument, List<Metric
                     }
 
                     if (this.instrumentIdentityToMetric.TryGetValue(instrumentIdentity, out var existingMetric))
-                    {
-                        metrics.Add(existingMetric);
-                        continue;
-                    }
-
-                    if (this.metricStreamNames.Contains(metricStreamName))
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricReaderExt.cs:141]: we still need this check, and warning, for the following test

```
[Fact]
        public void Test()
        {
            var exportedItems = new List<Metric>();

            using var meter = new Meter($"{Utils.GetCurrentMethodName()}");
            var meterProviderBuilder = Sdk.CreateMeterProviderBuilder()
                .AddMeter(meter.Name)
                .AddView("uncommon", MetricStreamConfiguration.Drop)
                .AddInMemoryExporter(exportedItems);

            using var meterProvider = meterProviderBuilder.Build();

            var instrument1 = meter.CreateCounter<long>("name", description: "desc1");
            var instrument2 = meter.CreateCounter<long>("name", description: "desc2");

            instrument1.Add(10);
            instrument2.Add(10);

            meterProvider.ForceFlush(MaxTimeToAllowForFlush);

            Assert.Equal(2, exportedItems.Count);
            Assert.Equal("name", exportedItems[0].Name);
            Assert.Equal("desc1", exportedItems[0].Description);
            Assert.Equal("name", exportedItems[1].Name);
            Assert.Equal("desc2", exportedItems[1].Description);
        }
```

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs (hunk 4 above)
Comment (alanwest) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs:155]: Yep good catch.

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs (hunk 2 above)
Comment (alanwest) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs]: Yep, agree, thanks

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs (hunk 7 above)
Comment (alanwest) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs]: On second thought, I removed this log message. I think it is unnecessary. It is not an issue of a duplicate conflicting metric instrument. The `if (metrics.Count == 0)` case is meant to handle instruments mapping to identical metric streams (i.e., same name, kind, description, etc) - so it is safe to aggregate together.

For example, this test covers this case

https://github.com/open-telemetry/opentelemetry-dotnet/blob/3bf47f835f2899e13df88f01a4870228db93df41/test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs#L503-L524

Both `instrument1` and `instrument2` match the first view, the metric stream is identical, so they can be aggregated together. The second view is ignored, and a log message is produced in this case.

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs (hunk 6 above)
Comment (alanwest) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs]: Moved and added a test for this scenario https://github.com/open-telemetry/opentelemetry-dotnet/pull/3129/commits/462d1afe028c897d34bcc35f86f9e7b211bb474b

**Comment by alanwest** [comment]
We've settled on #3006 as that is the direction the spec will be taking

**Pull request closed without merging**